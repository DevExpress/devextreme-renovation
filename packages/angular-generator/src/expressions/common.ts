import {
  Block, Call as BaseCall, capitalizeFirstLetter, Identifier, ReturnStatement, SimpleExpression, StringLiteral
} from '@devextreme-generator/core';
import { Function } from './functions/function';
import { toStringOptions } from '../types';

export class Call extends BaseCall {
  toString(options?: toStringOptions) {
    if (this.expression.toString() === 'this.props.hasOwnProperty') {
      return this.compileHasOwnProperty(options);
    }
    return super.toString(options);
  }

  compileHasOwnProperty(options?: toStringOptions) {
    let argument = this.arguments?.[0];
    if (argument instanceof StringLiteral) {
      const value = argument.valueOf();
      const nestedNames = options?.members.filter((m) => m.isNested).map((m) => m.name) ?? [];
      if (
        this.arguments.length > 0
        && nestedNames.some((name) => name === value.toString())
      ) {
        argument = new StringLiteral(`__${value}`, argument.quoteSymbol);
        // const functionName = `hasOwnNested${capitalizeFirstLetter(value)}`;
        // const hasOwnNestedFunction = new Function(
        //   undefined,
        //   undefined,
        //   '',
        //   new Identifier(functionName),
        //   undefined, [], undefined,
        //   new Block([new ReturnStatement(new SimpleExpression(`this.hasOwnProperty("__${value}")||this.${value}!==undefined`))], true),
        //   {}
        // );
        return `this.hasOwnProperty(${argument}) || this.${value} !== undefined`;
      }
    }

    return `this.hasOwnProperty(${argument})`;
  }

  // createHasOwnNested(options?: toStringOptions) {
  //   const argument = this.arguments?.[0];
  // }
}
