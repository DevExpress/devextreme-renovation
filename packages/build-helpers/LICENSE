DevExtreme Development Utils are released as MIT-licensed.
