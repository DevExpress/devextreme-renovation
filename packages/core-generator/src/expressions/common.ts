import { toStringOptions } from '../types';
import { SyntaxKind } from '../syntaxKind';
import { ExpressionWithExpression, Expression, SimpleExpression } from './base';
import { TypeExpression } from './type';
import { compileTypeParameters } from '../utils/string';
import { StringLiteral } from './literal';

function getIdentifierExpressionFromVariable(
  expression: Expression,
  options?: toStringOptions,
) {
  const stringValue = expression.toString();
  if (options?.variables && options.variables[stringValue]) {
    return options.variables[stringValue];
  }
  return undefined;
}

export class Identifier extends SimpleExpression {
  valueOf() {
    return this.expression;
  }

  toString(options?: toStringOptions) {
    const baseValue = super.toString();
    if (options?.variables && options.variables[baseValue]) {
      let expression = options.variables[baseValue];
      if (expression instanceof Paren) {
        expression = expression.expression;
      }
      if (options.variables[baseValue].toString() === baseValue) {
        return baseValue;
      }
      return options.variables[baseValue].toString(options);
    }
    return baseValue;
  }

  getDependency(options: toStringOptions) {
    const expression = getIdentifierExpressionFromVariable(this, options);
    if (expression) {
      return expression.getDependency(options);
    }
    return [];
  }
}

export class TypeOf extends ExpressionWithExpression {
  toString(options?: toStringOptions) {
    return `${SyntaxKind.TypeOfKeyword} ${this.expression.toString(options)}`;
  }
}

export class Void extends ExpressionWithExpression {
  toString(options?: toStringOptions) {
    return `${SyntaxKind.VoidKeyword} ${this.expression.toString(options)}`;
  }
}

export class Delete extends ExpressionWithExpression {
  toString(_options?: toStringOptions) {
    return `${SyntaxKind.DeleteKeyword} ${super.toString()}`;
  }
}

export class Paren extends ExpressionWithExpression {
  toString(options?: toStringOptions) {
    return `(${super.toString(options)})`;
  }
}

export class Call extends ExpressionWithExpression {
  constructor(
    expression: Expression,
    public typeArguments?: TypeExpression[],
    public argumentsArray: Expression[] = [],
  ) {
    super(expression);
  }

  compileTypeArguments() {
    return compileTypeParameters(this.typeArguments);
  }

  get arguments() {
    return this.argumentsArray.map((a) => a);
  }

  toString(options?: toStringOptions) {
    if (this.expression.toString() === 'this.props.hasOwnProperty') {
      const argument = this.arguments?.[0];
      if (argument instanceof StringLiteral) {
        const value = argument.valueOf();
        return this.compileHasOwnProperty(value, options);
      }
    }
    return this.compileDefaultToString(options);
  }

  getDependency(options: toStringOptions) {
    const argumentsDependency = this.argumentsArray.reduce((d: string[], a) => d.concat(a.getDependency(options)), []);
    return super.getDependency(options).concat(argumentsDependency);
  }

  compileHasOwnProperty(_value: string, options?: toStringOptions) {
    return this.compileDefaultToString(options);
  }

  compileDefaultToString(options?: toStringOptions) {
    return `${this.expression.toString(
      options,
    )}${this.compileTypeArguments()}(${this.argumentsArray
      .map((a) => a.toString(options))
      .join(',')})`;
  }
}

export class New extends Call {
  toString(options?: toStringOptions) {
    return `${SyntaxKind.NewKeyword} ${super.toString(options)}`;
  }
}

export class CallChain extends Call {
  questionDotToken: string;

  constructor(
    expression: Expression,
    questionDotToken = '',
    typeArguments: any,
    argumentsArray: Expression[] = [],
  ) {
    super(expression, typeArguments, argumentsArray);
    this.questionDotToken = questionDotToken;
  }

  toString(options?: toStringOptions) {
    return `${this.expression.toString(options)}${
      this.questionDotToken
      }(${this.argumentsArray.map((a) => a.toString(options)).join(',')})`;
  }
}

export class NonNullExpression extends ExpressionWithExpression {
  toString(options?: toStringOptions) {
    return `${super.toString(options).replace(/[?!]$/, '')}!`;
  }
}

export class AsExpression extends ExpressionWithExpression {
  type: TypeExpression;

  constructor(expression: Expression, type: TypeExpression) {
    super(expression);
    this.type = type;
  }

  toString(options?: toStringOptions) {
    return `${super.toString(options)} ${SyntaxKind.AsKeyword} ${this.type}`;
  }
}
