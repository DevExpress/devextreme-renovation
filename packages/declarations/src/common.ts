import * as react from 'react';
import { createPortal } from 'react-dom';

export type DefaultsRule<T> = {
  device: any;
  options: Partial<T>;
};

export const React = react;
export const Fragment = react.Fragment;

declare type PortalProps = {
  children: react.ReactNode;
  container: Element | null | undefined;
};
export const Portal: react.ElementType<PortalProps> = ({
  children,
  container,
}) => (container instanceof Element ? createPortal(children, container) : null);

export type ComponentParameters = {
  name?: string;
  components?: any[];
  /**
   * Function that receives a component viewModel and returns a view
   */
  view: Function;

  /**
   * Set to `null` to not generate 'defaultOptions' method
   */
  defaultOptionRules?: DefaultsRule<any>[] | null;

  /**
   * Set to `true` if component is SVG-based
   */
  isSVG?: boolean;

  /**
   * jQuery wrapper specific settings
   */
  jQuery?: {
    /**
     * If jQuery widget should be registered
     */
    register?: boolean;
    /**
     * Custom base widget for wrapper
     */
    component?: any;
  };
};

/**
 * Class Decorator.
 * Declare class that contains all internal logic and public methods.
 * Class should extend JSXComponent<ComponentBindings>.
 * Class can contain properties with the following decorators: [ForwardRef](https://github.com/DevExpress/devextreme-renovation/blob/master/guidelines.md#forwardref) [Ref](https://github.com/DevExpress/devextreme-renovation/blob/master/guidelines.md#ref), InternalState(optional), Provider, Consumer (see [Context](https://github.com/DevExpress/devextreme-renovation/blob/master/guidelines.md#context)).
 * Class can contain methods with the following decorators: [Effect](https://github.com/DevExpress/devextreme-renovation/blob/master/guidelines.md#effect), [Method](https://github.com/DevExpress/devextreme-renovation/blob/master/guidelines.md#method).
 * Class can contain methods without any decorator.
 */
export function Component(arg: ComponentParameters) {
  return function ComponentDecorator(constructor: Function) {
    constructor.prototype.render = function () {
      return arg.view(this);
    };
  };
}

/**
 * Class Decorator.
 * Use for definition of Component interface.
 * Declare class that contains is everything that user of component can pass to it.
 * Class can contains only properties with decorators OneWay, TwoWay, Slot, Template.
 */
export function ComponentBindings() {
  return function ComponentBindings(_constructor: Function) {};
}

const propertyDecorator = function (_target: any, _propertyKey: string) {};

/**
 * Property Decorator.
 * Define a property that user can pass to component. Component can't write to it.
 */
export const OneWay = () => propertyDecorator;

/**
 * Property Decorator.
 * Define a property that user can pass to component.
 */
export const ForwardRef = () => propertyDecorator;
/**
 * Property Decorator.
 * Define a property that user can pass to component. Component can write to it.
 * Any change of this property causes component re-render, also user gets notified about that change (in a different way, depending on target platform)
 */
export const TwoWay = (_args?: {
  /**
   * Specify whether to use this prop as model binding
   */
  isModel?: boolean;
}) => propertyDecorator;
/**
 * Property Decorator.
 * Define a property that user can pass to component as property or as nested component.
 */
export const Nested = () => propertyDecorator;

/**
 * Property Decorator.
 * Declares an input property that the user can set for custom rendering.
 * Property type should be a function that receives model and returns markup.
 */
export const Template = () => propertyDecorator;
/**
 * Property Decorator.
 * Declares an input property that the user can set custom rendering part of widget.
 */
export const Slot = (_args?: {
  /**
   * Specify whether to pass content as svg
   */
  isSVG?: boolean;
}) => propertyDecorator;
/**
 * Method Decorator.
 * Declare A method that be available for external using
 */
export const Method = () => propertyDecorator;
/**
 * Property Decorator.
 * Declares input property that is component event handler
 */
export const Event = (_args?: { actionConfig?: any }) => propertyDecorator;
/**
 * Property Decorator
 * Declares a widget property. Any change of it causes component re-render
 */
export const InternalState = () => propertyDecorator;
/**
 * Property Decorator
 * Declares a widget property. Any change of it doesn't causes component re-render and doesn't call effects.
 */
export const Mutable = () => propertyDecorator;

class Context {}

export const Provider = (_Context: Context) => propertyDecorator;

export const Consumer = (_Context: Context) => propertyDecorator;

export const createContext = <T>(_defaultValue: T) => new Context();

export const Listen = (
  _eventName?: string,
  _parameters?: { target?: Document | Window | string },
) => propertyDecorator;
/**
 * Property Decorator.
 * Declare a reference on DOM Element
 */
export const Ref = () => propertyDecorator;

export type RefObject<T = any> = {
  bivarianceHack(instance: (T & Element) | null): void;
}['bivarianceHack'] & { current: T | null };

/**
 * Method Decorator.
 * Declare a logic that should be executed after component is rendered.
 * All DOM references is available here.
 */
export const Effect = (_args?: { run?: 'once' | 'always' }) => propertyDecorator;

interface JSXComponentBase<_P> {
  context: any;
  forceUpdate: any;
  refs: any;
  render(): any;
  setState(): any;
  state(): any;
  isReactComponent: any;
}
class JSXComponentBase<_P> {}
/**
 * A function that returns base class for any Component.
 * Pass ComponentBindings as an argument
 */
export function JSXComponent<
  PropsType,
  RequiredProps extends keyof PropsType = Exclude<
  keyof PropsType,
  keyof PropsType
  >,
>(Props?: { new (): PropsType & { ref?: React.Component<PropsType> } }) {
  type DefaultPropsType = Omit<PropsType, RequiredProps> & {
    [name: string]: any;
  };
  type RealPropsType = Partial<Omit<PropsType, RequiredProps>> &
  Pick<PropsType, RequiredProps>;
  const BaseComponent = class extends JSXComponentBase<RealPropsType> {
    static defaultProps: DefaultPropsType = ((Props && new Props())
      || {}) as DefaultPropsType;

    static isReactComponent = {};

    constructor(props: RealPropsType) {
      super();
      this.props = props as PropsType;
    }

    props!: PropsType;

    restAttributes: { [name: string]: any } = {
      'rest-attributes': 'restAttributes',
    }; // for testing purpose
  };
  BaseComponent.prototype.isReactComponent = {}; // for enzyme tests
  return BaseComponent;
}

/**
 * JSX.Element with declared properties.
 */
export type JSXTemplate<
  PropsType = {},
  RequiredProps extends keyof PropsType = Exclude<
  keyof PropsType,
  keyof PropsType
  >,
> = React.JSXElementConstructor<
Partial<Omit<PropsType, RequiredProps>> &
Required<Pick<PropsType, RequiredProps>>
>;

/**
 * Type for styles.
 */
export type CSSAttributes = React.CSSProperties;

/**
 * Create Elements function
 * Import it for testing purpose
 */
export const h = react.createElement;
