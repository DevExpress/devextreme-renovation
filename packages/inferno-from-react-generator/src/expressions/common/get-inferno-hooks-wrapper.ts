import { ComponentInfo } from '../../component-info';

export function getInfernoHooksWrapper(
  { name, jQueryRegistered, hasApiMethod }: ComponentInfo,
  addExport = false,
  isPureComponent = false,
): string {
  if (hasApiMethod) {
    return `
    function Hooks${name}(props, ref) {
    return <${jQueryRegistered ? 'InfernoWrapperComponent' : 'HookContainer'} renderFn={
        React${name}
      } renderProps={props} renderRef={ref} ${isPureComponent ? 'pure' : ''}/>
    }
    const ${name} = forwardRef(Hooks${name});
    ${addExport ? `export { ${name} }` : ''}
  `;
  }
  return `
  function Hooks${name}(props) {
  return <${jQueryRegistered ? 'InfernoWrapperComponent' : 'HookContainer'} renderFn={
    ${name}
    } renderProps={props} ${isPureComponent ? 'pure' : ''}/>
  }
  export {Hooks${name} as ${name}}
  `;
}
