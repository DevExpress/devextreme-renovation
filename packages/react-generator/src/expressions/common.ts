import {
  Conditional,
  Identifier,
  New as BaseNew,
  toStringOptions,
  Call as BaseCall,
} from '@devextreme-generator/core';
import { PropertyAccess } from './property-access';

export class New extends BaseNew {
  toString(options?: toStringOptions): string {
    const componentInputs = options?.componentInputs || [];
    if (componentInputs.length) {
      const matchedInput = componentInputs.find(
        (c) => c.name === this.expression.toString(),
      );
      if (matchedInput?.members.some((m) => m.isNested)) {
        const conditional = new Conditional(
          new PropertyAccess(
            this.expression,
            new Identifier('__defaultNestedValues'),
          ),
          new PropertyAccess(
            this.expression,
            new Identifier('__defaultNestedValues'),
          ),
          this.expression,
        );
        return conditional.toString();
      }
      if (matchedInput) {
        return this.expression.toString();
      }
    }
    return super.toString(options);
  }
}

export class Call extends BaseCall {
  compileHasOwnProperty(value: string, options?: toStringOptions): string {
    const nestedAndTwoWayNames = options?.members.filter(
      (m) => m.isNested || m.decorators?.[0]?.toString() === '@TwoWay()',
    )
      .map((m) => m.name) ?? [];
    const isTwoWayOrNested = (nestedAndTwoWayNames.some((name) => name === value.toString()));
    if (
      this.expression instanceof PropertyAccess
      && this.arguments.length > 0
      && isTwoWayOrNested
    ) {
      const propsContext = options?.newComponentContext ? `${options?.newComponentContext}.` : '';
      return `${this.expression.expression.toString(options)}.${value} !== undefined || ${propsContext
      }props.hasOwnProperty("${value}")`;
    }
    return super.compileHasOwnProperty(value, options);
  }
}
