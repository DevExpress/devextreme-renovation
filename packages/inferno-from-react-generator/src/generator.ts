import BaseGenerator, {
  Binary,
  Call,
  Decorator, Expression, Identifier, ImportClause, ExportDeclaration,
  JsxAttribute, JsxClosingElement, JsxElement, JsxExpression,
  JsxOpeningElement, JsxSelfClosingElement, JsxSpreadAttribute, NamedImports,
  PropertyAccess, SimpleExpression, StringLiteral, VariableDeclarationList,
  VariableStatement, GeneratorResult,
} from '@devextreme-generator/core';
import { ImportDeclaration } from './expressions/import';

interface ComponentInfo {
  name: string;
  jQueryRegistered: boolean;
  hasApiMethod: boolean;
  isGenerated: boolean;
}
export class ReactInfernoGenerator extends BaseGenerator {
  private components: { [key: string]: ComponentInfo } = {};

  getPlatform(): string {
    return 'inferno';
  }

  compileHooksWrapper({ name, jQueryRegistered, hasApiMethod }: ComponentInfo): string {
    if (hasApiMethod) {
      return `
      function Hooks${name}(props, ref) {
      return <${jQueryRegistered ? 'InfernoWrapperComponent' : 'HookContainer'} renderFn={
          React${name}
        } renderProps={props} renderRef={ref}/>
      }
      const ${name} = forwardRef(Hooks${name});
      export { ${name} };
    `;
    }
    return `
    function Hooks${name}(props) {
    return <${jQueryRegistered ? 'InfernoWrapperComponent' : 'HookContainer'} renderFn={
      ${name}
      } renderProps={props}/>
    }
    export {Hooks${name} as ${name}}
    `;
  }

  createExpressionStatement(expression: Expression): Expression {
    const expresssionStatement = super.createExpressionStatement(expression);
    if ((expression instanceof Binary)
      && (expression.left instanceof PropertyAccess)) {
      const componentName = expression.left.expression.toString();
      const componentInfo = this.components[componentName];
      if (componentInfo && componentInfo.hasApiMethod) {
        const isdefaultProps = expression.left.name.toString() === 'defaultProps';
        if (isdefaultProps) {
          const expressionStr = expresssionStatement.toString();
          expresssionStatement.toString = () => (`${expressionStr};
            Hooks${componentName}.defaultProps = ${componentName}.defaultProps;`);
        }
      }
    }
    return expresssionStatement;
  }

  createComponentComments(comments: string[]): void {
    comments.forEach((comment) => {
      if (comment.indexOf('Component={') !== -1) {
        const componentInfoString = comment.slice(comment.indexOf('{'));
        const componentInfo: ComponentInfo = JSON.parse(componentInfoString);
        this.components[componentInfo.name] = componentInfo;
      }
    });
  }

  createImportDeclaration(
    decorators: Decorator[] | undefined,
    modifiers: string[] | undefined,
    importClause: ImportClause = new ImportClause(),
    moduleSpecifier: StringLiteral,
  ): ImportDeclaration {
    return new ImportDeclaration(decorators, modifiers, importClause,
      moduleSpecifier, this.getContext());
  }

  createJsxOpeningFragment(
    _: Expression,
    typeArguments: any[],
    attributes?: Array<JsxAttribute | JsxSpreadAttribute>,
  ): JsxOpeningElement {
    return new JsxOpeningElement(new SimpleExpression(''), typeArguments, attributes, this.getContext());
  }

  createJsxClosingFragment(): JsxClosingElement {
    return new JsxClosingElement(new SimpleExpression(''));
  }

  createJsxFragment(
    openingElement: JsxOpeningElement,
    children: (JsxElement | string | JsxExpression | JsxSelfClosingElement)[],
    closingElement: JsxClosingElement,
  ): JsxElement {
    return new JsxElement(openingElement, children, closingElement);
  }

  createExportDeclaration(
    decorators: Decorator[] | undefined,
    modifiers: string[] | undefined,
    exportClause: NamedImports | undefined,
    moduleSpecifier?: Expression,
  ): ExportDeclaration | string {
    if (exportClause) {
      const componentInfo = Object.entries(this.components)
        .find(([key]) => exportClause.has(key));
      if (componentInfo) {
        componentInfo[1].isGenerated = true;
        return `
      ${this.compileHooksWrapper(componentInfo[1])}`;
      }
    }
    return super.createExportDeclaration(decorators, modifiers, exportClause, moduleSpecifier);
  }

  createExportAssignment(
    _decorators: Decorator[] = [],
    _modifiers: string[] = [],
    _isExportEquals: any,
    expression: Expression,
  ): string {
    return `export default ${expression}`;
  }

  createVariableStatement(
    modifiers: string[] | undefined,
    declarationList: VariableDeclarationList,
  ): VariableStatement {
    const componentDeclaration = declarationList.declarations
      .find(({ name }) => this.components[name.toString()]);
    if (componentDeclaration) {
      componentDeclaration.name = new Identifier(`React${componentDeclaration.name.toString()}`);
      componentDeclaration.initializer = (componentDeclaration.initializer as Call)
        .argumentsArray[0];
    }
    return super.createVariableStatement(modifiers, declarationList);
  }

  postProcessResult(results: GeneratorResult[]): void {
    const hasNotProcessedComponents = Object.values(this.components)
      .filter(({ isGenerated }) => (!isGenerated));
    if (hasNotProcessedComponents.length > 0) {
      throw new Error(`Not all components were processed: ${hasNotProcessedComponents.map(({ name }) => (name))}`);
    }
    const hasAbsentDefaultExport = Object.values(this.components)
      .filter(({ name, jQueryRegistered }) => jQueryRegistered && results[0].code.indexOf(`export default ${name}`) === -1);
    if (hasAbsentDefaultExport.length > 0) {
      throw new Error(`Not all jQuery registred components were default export: ${hasAbsentDefaultExport.map(({ name }) => (name))}`);
    }
    this.components = {};
  }
}
