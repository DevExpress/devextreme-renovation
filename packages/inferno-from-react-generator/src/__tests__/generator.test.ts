/* eslint-disable import/no-extraneous-dependencies */
import { compile } from '@devextreme-generator/build-helpers';
import { resolve } from 'path';
import * as fs from 'fs';
import { InfernoFromReactGenerator } from '../generator';
import { createTestGenerator } from './common';

describe('generator', () => {
  beforeAll(() => {
    if (fs.readdirSync(`${__dirname}/react-test-cases`).length !== fs.readdirSync(`${__dirname}/test-cases`).length) {
      throw new Error('Test cases are not up to date. Run `npm run update:testcases`');
    }
    compile(
      `${__dirname}/test-cases/`,
      `${__dirname}/componentFactory/`,
      'inferno',
      /.js(x?)$/,
    );
  });

  function createGenerator(): InfernoFromReactGenerator {
    const root = './test-cases/';
    const generator = new InfernoFromReactGenerator();
    generator.setContext(
      {
        dirname: resolve(root),
        path: resolve(root),
      },
    );
    return generator;
  }

  it('adds import for HookContainer', async () => {
    const generator = await createTestGenerator('import-for-hook-container', createGenerator());
    expect(generator[0].code).toMatchSnapshot();
  });

  it('adds import for InfernoWrapperComponent', async () => {
    const generator = await createTestGenerator('import-for-inferno-wrapper-component', createGenerator());
    expect(generator[0].code).toMatchSnapshot();
  });

  it('does not remove other imports from react', async () => {
    const generator = await createTestGenerator('import-without-react', createGenerator());
    expect(generator[0].code).toMatchSnapshot();
  });

  it('removes React namespace import', async () => {
    const generator = await createTestGenerator('import-react-namespace', createGenerator());
    expect(generator[0].code).toMatchSnapshot();
  });

  it('removes React default import', async () => {
    const generator = await createTestGenerator('import-react-default', createGenerator());
    expect(generator[0].code).toMatchSnapshot();
  });

  it('replace fragments', async () => {
    const generator = await createTestGenerator('fragments', createGenerator());
    expect(generator[0].code).toMatchSnapshot();
  });

  it('simple component only name', async () => {
    const generator = await createTestGenerator('simple-component', createGenerator());
    expect(generator[0].code).toMatchSnapshot();
  });

  it('forward-ref and useImperativeHandler', async () => {
    const generator = await createTestGenerator('forward-ref', createGenerator());
    expect(generator[0].code).toMatchSnapshot();
  });

  it('forward-ref and useImperativeHandler direct export const', async () => {
    const generator = await createTestGenerator('forward-ref-direct-export', createGenerator());
    expect(generator[0].code).toMatchSnapshot();
  });

  it('jQueryRegistered', async () => {
    const generator = await createTestGenerator('jquery-registred', createGenerator());
    expect(generator[0].code).toMatchSnapshot();
  });

  it('file contains component content but doesnt have "import {SimpleComponent}" statement', async () => {
    await expect(async () => createTestGenerator('no-component-export', createGenerator()))
      .rejects
      .toEqual(new Error('Not all components were processed: SimpleComponent'));
  });

  it('render several components should not raise error', async () => {
    const generator = createGenerator();
    await createTestGenerator('jquery-registred', generator);
    await createTestGenerator('jquery-registred2', generator);
  });

  it('dont change "export {}"', async () => {
    const generator = await createTestGenerator('types', createGenerator());
    expect(`${generator[0].code}`).toMatchSnapshot();
  });

  it('React.memo', async () => {
    const generator = await createTestGenerator('react-memo', createGenerator());
    expect(`${generator[0].code}`).toMatchSnapshot();
  });
});
