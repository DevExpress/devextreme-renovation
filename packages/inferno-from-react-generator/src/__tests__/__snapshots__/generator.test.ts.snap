// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generator forward-ref and useImperativeHandler 1`] = `
"import {
  forwardRef,
  useImperativeHandle,
  HookContainer,
} from '@devextreme/runtime/inferno-hooks';
export const viewFunction = () => <div />;
const ReactForwardRef = (_, ref) => {
  useImperativeHandle(ref, () => ({ focus: () => {} }), []);
  return viewFunction();
};

function HooksForwardRef(props, ref) {
  return (
    <HookContainer
      renderFn={ReactForwardRef}
      renderProps={props}
      renderRef={ref}
    />
  );
}
const ForwardRef = forwardRef(HooksForwardRef);
export { ForwardRef };
export default ForwardRef;
ForwardRef.defaultProps = {};
HooksForwardRef.defaultProps = ForwardRef.defaultProps;
"
`;

exports[`generator jQueryRegistered 1`] = `
"function Component1(props) {
  return 'content';
}

function HooksComponent1(props) {
  return <InfernoWrapperComponent renderFn={Component1} renderProps={props} />;
}
export { HooksComponent1 as Component1 };
export default Component1;
"
`;

exports[`generator replace fragments 1`] = `
"export const viewFunction = (viewModel) => (
  <>
    <div>a</div>
  </>
);
function SimpleComponent(props) {
  return viewFunction({ props: { ...props } });
}

function HooksSimpleComponent(props) {
  return <HookContainer renderFn={SimpleComponent} renderProps={props} />;
}
export { HooksSimpleComponent as SimpleComponent };
"
`;

exports[`generator replace imports 1`] = `
"import { getTemplate } from '@devextreme/runtime/inferno-hooks';
import {
  useCallback,
  HookContainer,
} from '@devextreme/runtime/inferno-hooks';
export const viewFunction = (m) => (
  <>
    <div>{getTemplate('test')}</div>
  </>
);
function Icon(props) {
  const __sourceType = useCallback(() => '', []);
  return viewFunction({});
}

function HooksIcon(props) {
  return <HookContainer renderFn={Icon} renderProps={props} />;
}
export { HooksIcon as Icon };
export default Icon;
"
`;
